go/teleportr/drivers/disburser/driver.go

```diff
@@ -0,0 +1,518 @@
+			Addr:   deposit.Address,
+		}
+		disbursements = append(disbursements, disbursement)
+		depositIDs = append(depositIDs, deposit.ID)
+	}
+
+	log.Info(name+" crafting batch tx", "start", start, "end", end,
+		"nonce", nonce)
+
+	d.metrics.NumElementsPerBatch.Observe(float64(len(disbursements)))
+
+	log.Info(name+" batch constructed", "num_disbursements", len(disbursements))
+
+	gasPrice, err := d.cfg.L1Client.SuggestGasPrice(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	opts, err := bind.NewKeyedTransactorWithChainID(
+		d.cfg.PrivKey, d.cfg.ChainID,
+	)
+	if err != nil {
+		return nil, err
+	}
+	opts.Context = ctx
+	opts.Nonce = nonce
+	opts.GasPrice = gasPrice
+	opts.NoSend = true
+
+	tx, err := d.disburserContract.Disburse(opts, start, disbursements)
+	if err != nil {
+		return nil, err
+	}
+
+	d.currentDepositIDs = depositIDs
+
+	return tx, nil
+}
+
+// UpdateGasPrice signs an otherwise identical txn to the one provided but with
+// updated gas prices sampled from the existing network conditions.
+//
+// NOTE: Thie method SHOULD NOT publish the resulting transaction.
+func (d *Driver) UpdateGasPrice(
+	ctx context.Context,
+	tx *types.Transaction,
+) (*types.Transaction, error) {
+
+	gasPrice, err := d.cfg.L1Client.SuggestGasPrice(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	opts, err := bind.NewKeyedTransactorWithChainID(
+		d.cfg.PrivKey, d.cfg.ChainID,
+	)
+	if err != nil {
+		return nil, err
+	}
+	opts.Context = ctx
+	opts.Nonce = new(big.Int).SetUint64(tx.Nonce())
+	opts.GasPrice = gasPrice
+	opts.NoSend = true
+
+	return d.rawDisburserContract.RawTransact(opts, tx.Data())
+}
+
+// SendTransaction injects a signed transaction into the pending pool for
+// execution.
+func (d *Driver) SendTransaction(
+	ctx context.Context,
+	tx *types.Transaction,
+) error {
+	return d.cfg.L2Client.SendTransaction(ctx, tx)
+}
+
+// OnConfirmation executes after a transaction has sufficiently confirmed,
+// passing in the resulting transaction receipt. This is used to perform any
+// additional bookkeeping or cleanup regarding the last transaction.
+func (d *Driver) OnConfirmation(
+	ctx context.Context,
+	receipt *types.Receipt,
+) error {
+
+	depositIDs := d.currentDepositIDs
+	d.currentDepositIDs = nil
+
+	var blockTimestamp time.Time
+	header, err := d.cfg.L2Client.HeaderByNumber(ctx, receipt.BlockNumber)
+	if err != nil {
+		log.Warn("Unable to fetch block by number",
+			"blockNumber", receipt.BlockNumber,
+			"err", err)
+
+		// Not super critical that this is the actual timestamp, can always
+		// rectify later. Not worth failing over.
+		blockTimestamp = time.Unix(time.Now().Unix(), 0)
+	} else {
+		blockTimestamp = time.Unix(int64(header.Time), 0)
+	}
+
+	for _, depositID := range depositIDs {
+		log.Info("Marking disbursement success",
+			"depositId", depositID,
+			"txHash", receipt.TxHash,
+			"blockNumber", receipt.BlockNumber,
+			"blockTimestamp", blockTimestamp)
+
+		err = d.cfg.Database.UpsertDisbursement(
```

Comment:
If a disbursement fails to be marked as successful, what happens?

Reply:
it would show up in the db as confirmed but not disbursed, and would require manual intervention to add the disbursement. the best way to get around this IMO is to add a database that tracks `txhash => (startId, endId)` and then we can recover after startups or in `GetBatchBlockRange`. this is a little more involved though, as we would need a call back from within the `txmgr` before it publishes, but it's definitely the safer route

Reply:
ah, actually we can just do the tracking in the `SendTransaction` iface method 😎